### ISSUE TITLE
MQTT client disconnecting while continuously publishing

### ISSUE THREAD

**Issue by miks** [eclipse-paho/paho.mqtt.c#609]
I'm currently doing refactoring by moving away from mosquitto to paho as there is no websocket support in mosquitto library. Library used in my case is paho-cpp with async client, but I believe problems lies somewhere in paho-c.

mqtt keepalive: 5 seconds
broker: EMQ at localhost 
connection: websockets (no SSL)
paho-c version: v1.3.0

```
20190130 201356.879 5 MacBook-Pro.local -> PINGREQ (0)
20190130 201356.881 5 MacBook-Pro.local <- PINGRESP
20190130 201358.210 5 MacBook-Pro.local -> PUBLISH qos: 0 retained: 1 (0)
20190130 201358.211 5 MacBook-Pro.local -> PUBLISH qos: 0 retained: 1 (0)
20190130 201358.211 5 MacBook-Pro.local -> PUBLISH qos: 0 retained: 1 (0)
20190130 201358.211 5 MacBook-Pro.local -> PUBLISH qos: 0 retained: 1 (0)
20190130 201402.000 5 MacBook-Pro.local -> PINGREQ (0)
20190130 201403.999 PINGRESP not received in keepalive interval for client MacBook-Pro.local on socket 5, disconnecting
20190130 201404.000 5 MacBook-Pro.local -> DISCONNECT (0)
```

When sending same data with mosquitto library no disconnection happens at all.
Also, when no data is published there is no missing PINGRESP.

**Comment by miks** [comment]
There is very similar issue #570, but if I understand correctly problem described already has been solved in 1.3.0

**Comment by icraggs** [comment]
Looks more likely to be related to #590.  5 seconds is a lower keepalive interval than I expected anyone to use.

**Comment by miks** [comment]
I'm getting the same problem also with 10 seconds.
As I already mentioned I'm using paho-cpp async client which goes something like this:
```
  while(run) {
    while(!publish_queue.empty()) {
        MqttMessage message = publish_queue.front();
        mqtt::message_ptr pubmsg = mqtt::make_message(message.topic, message.payload.c_str(), MQTT_QOS, message.persisted);
        mqtt::delivery_token_ptr pubtok = client->publish(pubmsg);
        pubtok->wait();
        publish_queue.pop();
    }

    usleep(QUEUE_READ_INTERVAL);
  }
```
where `publish_queue` is queue which get populated from another thread every 40 or so seconds.
As I have problems only when something gets published, can it be something related to this statement?:

```
This API is not thread safe however - it is not possible to call it from multiple
threads without synchronization. You can use the MQTTAsync API for that.
```

Another strange fact is that those publish events actually is not delivered at broker side. I can see all previous publish events except the ones before last PINGREQ and DISCONNECT.

**Comment by miks** [comment]
Switched to tcp:// instead of ws:/  and no disconnects anymore.
Afterward I tested also against eclipse-mosquitto and latest EMQ (v3) websockets and same error happens.
Maybe it's another bug, not related to one with timeouts.
Can it be useful to include log with MQTT_C_CLIENT_TRACE_LEVEL=MINIMUM or more?

**Comment by miks** [comment]
I can 100% confirm that I'm experiencing this "disconnection" issue only when using websockets.
Here is full log file: [mqtt.log.gz](https://github.com/eclipse/paho.mqtt.c/files/2818078/mqtt.log.gz)

From log file seems that this (starting around line #3199) fragment is repeating (hanging?) and most likely blocking other calls:
```
20190131 174859.431 CONNECT sent, *rc is -22
20190131 174859.431 (41795584)   (2)> MQTTAsync_retry:3495
20190131 174859.431 (41795584)    (3)> MQTTProtocol_retry:730
20190131 174859.431 (41795584)    (3)< MQTTProtocol_retry:749
20190131 174859.431 (41795584)   (2)< MQTTAsync_retry:3505
20190131 174859.431 (41795584)  (1)< MQTTAsync_cycle:3818 (-22)
20190131 174859.431 (41795584)  (1)> MQTTAsync_cycle:3662
20190131 174859.431 (41795584)   (2)> Socket_getReadySocket:237
20190131 174859.431 (41795584)    (3)> Socket_continueWrites:928
20190131 174859.431 (41795584)    (3)< Socket_continueWrites:952 (0)
20190131 174859.431 (41795584)    (3)> isReady:213
20190131 174859.431 (41795584)    (3)< isReady:218 (1)
20190131 174859.431 (41795584)   (2)< Socket_getReadySocket:309 (5)
20190131 174859.431 m->c->connect_state = 0
20190131 174859.431 (41795584)   (2)> MQTTPacket_Factory:111
20190131 174859.431 (41795584)    (3)> WebSocket_getch:469
20190131 174859.431 (41795584)     (4)> WebSocket_receiveFrame:734
20190131 174859.431 (41795584)      (5)> Socket_getdata:363
20190131 174859.431 (41795584)       (6)> SocketBuffer_getQueuedData:163
20190131 174859.431 (41795584)       (6)< SocketBuffer_getQueuedData:188
20190131 174859.431 (41795584)      (5)< Socket_getdata:397
20190131 174859.431 (41795584)     (4)< WebSocket_receiveFrame:884 (-22)
20190131 174859.431 (41795584)    (3)< WebSocket_getch:506 (-22)
20190131 174859.431 (41795584)   (2)< MQTTPacket_Factory:164 (-22)
20190131 174859.431 m->c->connect_state = 0
```

I can also repeat this error against ws://broker.hivemq.com:8000/mqtt

**Comment by lt-holman** [comment]
Seems to disconnect on Websocket Pings according to the original message... I put in a fix a few months ago that was accepted for pings.  Try running on the develop branch build to see if the problem persists.

**Comment by lt-holman** [comment]
Test MQTT broker over websockets doesn't send websocket pings (as they are optional, according to the spec).  So the problem won't be seen there if there is a ping/pong problem.

**Comment by icraggs** [comment]
Ah, so websocket pings as well as MQTT.  One or other is superfluous!  I'll look to add that to the test server.  I did use the develop branch against HiveMQ when I saw this issue as reported.

**Comment by icraggs** [comment]
Well that's strange.  I added a few printfs to WebSocket_getRawSocketData, WebSocket_receiveFrame, and Socket_getdata, then used paho-c-sub and paho-c-pub against HiveMQ.  The subscriber seems to work whereas the publisher gets this problem - but it does have a problem with HiveMQ with the non-websocket connection too.  Not sure what's going on now.

**Comment by jumoog** [comment]
Eclipse or IBM has Docker/Kubernetes Server?

Maybe its a good idea to test against a local Server and check Server logs.

See: https://www.hivemq.com/downloads/docker/

**Comment by piotrwach** [comment]
I have seen the same problem, and fixed it. Short version is that code for reading websocket frames doesn't handle properly case when it is not able to read the whole frame in one cycle (`MQTTAsync_cycle`).

I want to submit this as a PR, but I am waiting for a green light from my company.

**Comment by piotrwach** [comment]
For example, have a look at this line here: https://github.com/eclipse/paho.mqtt.c/blob/master/src/WebSocket.c#L825 - if it is not able to read the whole payload, then it will receive TCP_INTERRUPTED, and in the next cycle will try to read websocket frame again, but the websocket frame was already read in the previous cycle, so won't be there anymore.

**Comment by icraggs** [comment]
@piotrwach Any idea how long it might take for you to get permission?

**Comment by piotrwach** [comment]
@icraggs Probably next week.

**Comment by icraggs** [comment]
@piotrwach That would be great!

@jumoog I tried the HiveMQ docker image but it doesn't have websockets support out of the box.  The non-ws connection worked (locally), whereas when I connect to the public test server, the paho publisher sample gets disconnected regularly whether ws or not.  It looks like the broker is deliberately disconnecting the application, but I've no idea why.

**Comment by miks** [comment]
@piotrwach: any updates regarding your patch?

**Comment by piotrwach** [comment]
@icraggs @miks Nothing so far, but I am trying to get it done.

**Comment by miks** [comment]
@piotrwach please, give some hint whether something will come from your side or not.
I want to switch our software from using tcp to websocket (over SSL), but as currently websocket handling within paho is actually broken this is kind of stopper.

**Comment by piotrwach** [comment]
@miks I am working on it to get it done, but the fix is simple, as I already explained before, so if you can't wait, just do it yourself - it have fixed it within one day, so it doesn't take much time :)

**Comment by miks** [comment]
I will wait then, as I don't want to introduce some off-branch repositories. Thanks.

**Comment by piotrwach** [comment]
My ticket to contribute to this library has been approved. I am going to send pull request this week :)

**Comment by miks** [comment]
Right on time as I was planning to contribute this by myself :)

**Comment by icraggs** [comment]
I merged @piotrwach 's fix into the develop branch.  If anyone does verify it, please let us know.  Thanks.

**Comment by icraggs** [comment]
And thanks @piotrwach !

**Comment by Christoph15072019** [comment]
@icraggs 

I have tested the current development branch and again the connection to websocket was disconnected after some seconds. 

What I recognized is:
* WebSocketReceiver: the receiver thread doesn't check if the input socket has available data. The constructor of WebSocketFrame is called always. 
* on my system input.available() always returned 0. I am not sure if .available() really returns the amount of data ready to read.
* in case the constructor of WebSocketFrame is called without available data, byte firstByte = (byte) input.read(); throws a SocketTimeoutException 
* this exception results in stopping the receiver thread, closing the socket.

I did some changes and now it is working on my side. Are you interested in the updated source files.
